--- a/arch/riscv/dts/sun20i-d1.dtsi	2024-08-10 08:16:17.332222038 +0800
+++ b/arch/riscv/dts/sun20i-d1.dtsi	2024-08-10 08:12:30.869886489 +0800
@@ -35,6 +35,9 @@
 			i-cache-size = <32768>;
 			mmu-type = "riscv,sv39";
 			riscv,isa = "rv64imafdc";
+			riscv,isa-base = "rv64i";
+			riscv,isa-extensions = "i", "m", "a", "f", "d", "c", "zicntr", "zicsr",
+					       "zifencei", "zihpm";
 			#cooling-cells = <2>;
 
 			cpu0_intc: interrupt-controller {
