--- a/arch/riscv/dts/sun20i-d1.dtsi	2024-08-28 10:34:54.242186804 +0800
+++ b/arch/riscv/dts/sun20i-d1.dtsi	2024-08-28 21:24:42.154255833 +0800
@@ -35,6 +35,10 @@
 			i-cache-size = <32768>;
 			mmu-type = "riscv,sv39";
 			riscv,isa = "rv64imafdc";
+			riscv,isa-base = "rv64i";
+			riscv,isa-extensions = "i", "m", "a", "f", "d", "c", "zicntr", "zicsr",
+					       "zifencei", "zihpm", "xtheadvector";
+			thead,vlenb = <128>;
 			#cooling-cells = <2>;
 
 			cpu0_intc: interrupt-controller {
